core/drivers/versal_pm.c

```diff
@@ -0,0 +1,239 @@
+#define PM_CLOCK_DISABLE		37
+#define PM_CLOCK_GETSTATE		38
+#define PM_CLOCK_SETDIVIDER		39
+#define PM_CLOCK_GETDIVIDER		40
+#define PM_CLOCK_SETRATE		41
+#define PM_CLOCK_GETRATE		42
+#define PM_CLOCK_SETPARENT		43
+#define PM_CLOCK_GETPARENT		44
+#define PM_PLL_SET_PARAMETER		48
+#define PM_PLL_GET_PARAMETER		49
+#define PM_PLL_SET_MODE			50
+#define PM_PLL_GET_MODE			51
+#define PM_FEATURE_CHECK		63
+
+/* Loader API id */
+#define PM_LOAD_PDI			0x701
+
+/* PDI sources */
+#define PDI_SRC_JTAG		0x0
+#define PDI_SRC_QSPI24		0x1
+#define PDI_SRC_QSPI32		0x2
+#define PDI_SRC_SD0		0x3
+#define PDI_SRC_EMMC0		0x4
+#define PDI_SRC_SD1		0x5
+#define PDI_SRC_EMMC1		0x6
+#define PDI_SRC_USB		0x7
+#define PDI_SRC_OSPI		0x8
+#define PDI_SRC_SBI		0x9
+#define PDI_SRC_SMAP		0xA
+#define PDI_SRC_PCIE		0xB
+#define PDI_SRC_SD1_LS		0xE
+#define PDI_SRC_DDR		0xF
+
+struct versal_sip_payload {
+	uint32_t data[PAYLOAD_ARG_CNT];
+};
+
+const TEE_UUID versal_sip_uuid =  { 0x2ab9e4ec, 0x93b9, 0x11e7, {
+		0xa0, 0x19, 0xdf, 0xe0, 0xdb, 0xad, 0x0a, 0xe0 } };
+
+static uint32_t versal_sip_call(uint32_t pm_api_id, uint32_t arg0,
+				uint32_t arg1, uint32_t arg2, uint32_t arg3,
+				struct versal_sip_payload *payload)
+{
+	struct thread_smc_args args = {
+		.a0 = pm_api_id > PM_LOAD_PDI ?
+		      pm_api_id : PM_SIP_SVC | pm_api_id,
+		.a1 = reg_pair_to_64(arg1, arg0),
+		.a2 = reg_pair_to_64(arg3, arg2),
+	};
+
+	thread_smccc(&args);
+
+	if (payload) {
+		reg_pair_from_64(args.a0, &payload->data[1], &payload->data[0]);
+		reg_pair_from_64(args.a1, &payload->data[3], &payload->data[2]);
+		reg_pair_from_64(args.a2, &payload->data[5], &payload->data[4]);
+		reg_pair_from_64(args.a3, &payload->data[7], &payload->data[6]);
+	}
+
+	/* allow the PLM to output its debug information */
+	if (IS_ENABLED(CFG_VERSAL_TRACE_PLM))
+		mdelay(500);
+
+	return args.a0;
+}
+
+/* SIP call to program the FPGA has been obsoleted, use the PLM */
+TEE_Result versal_write_fpga(paddr_t pa)
+{
+	struct ipi_cmd cmd = { };
+
+	cmd.data[0] = PM_LOAD_PDI;
+	cmd.data[1] = PDI_SRC_DDR;
+	cmd.data[2] = pa >> 32;
+	cmd.data[3] = pa;
+	if (versal_mbox_notify(&cmd, NULL, NULL))
+		return TEE_ERROR_GENERIC;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result versal_soc_version(uint8_t *version)
+{
+	struct versal_sip_payload payload = { };
+	const uint32_t version_shift = 12;
+	uint32_t res = 0;
+
+	if (!version)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = versal_sip_call(PM_GET_CHIPID, 0, 0, 0, 0, &payload);
+	if (res) {
+		EMSG("Failed to retrieve version");
+		return TEE_ERROR_GENERIC;
+	}
+
+	*version = payload.data[2] >> version_shift;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result check_sip_uuid(struct versal_sip_payload *p)
+{
+	uint32_t time_low = TEE_U32_BSWAP(p->data[0]);
+	uint16_t time_mid = TEE_U32_BSWAP(p->data[2]) >> 16;
+	uint16_t time_hi  = TEE_U32_BSWAP(p->data[2]);
+	uint8_t uuid[16] = { };
+	size_t i = 0;
+	size_t j = 0;
+
+	for (i = 0; i < sizeof(time_low) && j < sizeof(uuid); i++, j++)
```

Comment:
You could simplify this by just checking that the four 32-bit words hold the expected values.
See for example optee_msg_api_uid_is_optee_api() at:

https://github.com/torvalds/linux/blob/2bca25eaeba6190efbfcb38ed169bd7ee43b5aaf/drivers/tee/optee/smc_abi.c#L1140-L1150

You should probably use `tee_uuid_from_octets()` instead if you want to convert this into a TEE_UUID.